samza-core/src/main/scala/org/apache/samza/system/SSPMetadataCache.java

```diff
@@ -0,0 +1,124 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.system;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.samza.util.Clock;
+
+
+/**
+ * Fetches and caches metadata about a set of SSPs. This can prefetch the metadata for a defined set of SSPs when it is
+ * fetching metadata for some arbitrary SSP.
+ */
+public class SSPMetadataCache {
+  private final SystemAdmins systemAdmins;
+  private final Duration cacheTTL;
+  private final Clock clock;
+  private final Set<SystemStreamPartition> sspsToPrefetch;
+
+  private final Object metadataRefreshLock;
+  private final ConcurrentHashMap<SystemStreamPartition, CacheEntry> cache;
+
+  public SSPMetadataCache(SystemAdmins systemAdmins, Duration cacheTTL, Clock clock,
+      Set<SystemStreamPartition> sspsToPrefetch) {
+    this.systemAdmins = systemAdmins;
+    this.cacheTTL = cacheTTL;
+    this.clock = clock;
+    this.sspsToPrefetch = sspsToPrefetch;
+
+    this.metadataRefreshLock = new Object();
+    this.cache = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Gets the metadata for an SSP. This will return a cached value if it is fresh enough. Otherwise, it will fetch the
+   * metadata from a source-of-truth.
+   * If the metadata for the SSP needs to be fetched, then this will also prefetch and cache the metadata for any
+   * sspsToPrefetch if they can be included in the same fetch call (e.g. same system) and if the corresponding metadata
+   * entries in the cache are stale.
+   *
+   * @param ssp SSP for which to get metadata
+   * @return metadata for the SSP; null if the source-of-truth returned no metadata
+   * @throws RuntimeException if there was an error in fetching metadata
+   */
+  public SystemStreamMetadata.SystemStreamPartitionMetadata getMetadata(SystemStreamPartition ssp) {
+    refreshMetadataIfNecessary(ssp);
+    CacheEntry cacheEntry = cache.get(ssp);
+    // shouldn't get null cacheEntry once refresh is done, but check anyways to be safe
+    return cacheEntry == null ? null : cacheEntry.getMetadata();
+  }
+
+  private void refreshMetadataIfNecessary(SystemStreamPartition requestedSSP) {
```

Comment:
s/refreshMetadataIfNecessary/maybeRefreshMetadata so that it's consistent with the rest of the Samza codebase?

Reply:
Resolved